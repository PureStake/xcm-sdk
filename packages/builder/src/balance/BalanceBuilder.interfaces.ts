import { ChainAssetId } from '@moonbeam-network/xcm-types';
import { Struct, u128 } from '@polkadot/types';
import { SubstrateQueryConfig } from '../types/substrate/SubstrateQueryConfig';

export interface BalanceConfigBuilder {
  build: (params: BalanceConfigBuilderPrams) => SubstrateQueryConfig;
}

export interface BalanceConfigBuilderPrams {
  address: string;
  asset: ChainAssetId;
}

export interface PalletBalancesAccountDataOld extends Struct {
  readonly free: u128;
  readonly reserved: u128;
  readonly miscFrozen: u128;
  readonly feeFrozen: u128;
}

export interface TokensPalletAccountData {
  free: u128;
  reserved: u128;
  frozen: u128;
}

export type EquilibriumSystemBalanceData = Array<
  [number, { positive: number }]
>;
