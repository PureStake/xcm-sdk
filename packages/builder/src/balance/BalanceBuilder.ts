/* eslint-disable sort-keys */
/* eslint-disable @typescript-eslint/no-use-before-define */

import { u128 } from '@polkadot/types';
import { PalletBalancesAccountData } from '@polkadot/types/lookup';
import { QueryConfig } from '../QueryConfig';
import {
  BalanceConfigBuilder,
  EquilibriumSystemBalanceData,
  PalletBalancesAccountDataOld,
  TokensPalletAccountData,
} from './BalanceBuilder.interfaces';

export function BalanceBuilder() {
  return {
    assets,
    system,
    tokens,
  };
}

function assets() {
  return {
    account: (): BalanceConfigBuilder => ({
      build: ({ address, asset }) =>
        new QueryConfig({
          module: 'assets',
          func: 'account',
          args: [asset, address],
          transform: (response: any): bigint =>
            (response.balance as u128).toBigInt(),
        }),
    }),
  };
}

function system() {
  return {
    account: (): BalanceConfigBuilder => ({
      build: ({ address }) =>
        new QueryConfig({
          module: 'system',
          func: 'account',
          args: [address],
          transform: (response: any): bigint => {
            const balance = response.data as PalletBalancesAccountData &
              PalletBalancesAccountDataOld;
            const frozen = balance.miscFrozen ?? balance.frozen;

            return BigInt(balance.free.sub(frozen).toString());
          },
        }),
    }),
    accountEquilibrium: (): BalanceConfigBuilder => ({
      build: ({ address, asset }) =>
        new QueryConfig({
          module: 'system',
          func: 'account',
          args: [address],
          transform: (response): bigint => {
            if (response.data.isEmpty) {
              return 0n;
            }

            const res = response.data.toJSON() as unknown;
            let balances: EquilibriumSystemBalanceData | undefined;

            if (Array.isArray(res)) {
              balances = res;
            }

            if (Array.isArray((res as any)?.v0?.balance)) {
              balances = (res as any).v0.balance;
            }

            if (!balances) {
              throw new Error("Can't get balance from Equilibrium chain");
            }

            const balance = balances.find(([assetId]) => assetId === asset);

            if (!balance) {
              return 0n;
            }

            return BigInt(balance[1].positive);
          },
        }),
    }),
  };
}

function tokens() {
  return {
    accounts: (): BalanceConfigBuilder => ({
      build: ({ address, asset }) =>
        new QueryConfig({
          module: 'tokens',
          func: 'accounts',
          args: [address, asset],
          transform: ({ free, frozen }: TokensPalletAccountData): bigint =>
            BigInt(free.sub(frozen).toString()),
        }),
    }),
  };
}
