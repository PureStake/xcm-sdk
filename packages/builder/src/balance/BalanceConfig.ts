import { CallConfigWithTransform } from '../CallConfigWithTransform';

export class BalanceConfig extends CallConfigWithTransform {}
