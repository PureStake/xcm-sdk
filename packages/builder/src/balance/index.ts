export * from './BalanceBuilder';
export * from './BalanceBuilder.interfaces';
