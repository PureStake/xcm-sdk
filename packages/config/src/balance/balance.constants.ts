export enum BalancePallet {
  Assets = 'assets',
  System = 'system',
  Tokens = 'tokens',
}

export enum BalanceFunction {
  Account = 'account',
  Accounts = 'accounts',
  Asset = 'asset',
}
