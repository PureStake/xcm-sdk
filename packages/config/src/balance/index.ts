export * from './balance';
export * from './balance.constants';
export * from './balance.interfaces';
