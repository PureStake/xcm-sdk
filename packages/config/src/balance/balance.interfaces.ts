import { u128 } from '@polkadot/types';
import { PalletBalancesAccountData } from '@polkadot/types/lookup';
import { AssetSymbol } from '../constants';
import { BalanceFunction, BalancePallet } from './balance.constants';

export type BalanceConfig<Symbols extends AssetSymbol = AssetSymbol> =
  | SystemBalanceConfig
  | AssetsBalanceConfig
  | TokensBalanceConfig<Symbols>;

export interface SystemBalanceConfig {
  pallet: BalancePallet.System;
  function: BalanceFunction.Account;
  path: ['data'];
  getParams: (account: string) => [string];
  calc: (data: PalletBalancesAccountData) => bigint;
}

export interface AssetsBalanceConfig {
  pallet: BalancePallet.Assets;
  function: BalanceFunction.Account;
  path: ['balance'];
  getParams: (account: string) => [number | bigint, string];
  calc: (balance: u128) => bigint;
}

export interface TokensPalletAccountData {
  free: u128;
  reserved: u128;
  frozen: u128;
}

export interface TokensBalanceConfig<
  Symbols extends AssetSymbol = AssetSymbol,
> {
  pallet: BalancePallet.Tokens;
  function: BalanceFunction.Accounts;
  path: [];
  getParams: (account: string) => [
    string,
    (
      | {
          Token: Symbols | 'MOVR' | 'KUSD';
        }
      | { ForeignAsset: number }
    ),
  ];
  calc: (data: TokensPalletAccountData) => bigint;
}

export interface MinBalanceConfig {
  pallet: BalancePallet.Assets;
  function: BalanceFunction.Asset;
  path: ['minBalance'];
  params: [number];
}
