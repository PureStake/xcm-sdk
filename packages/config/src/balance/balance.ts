import { PalletBalancesAccountData } from '@polkadot/types/lookup';
import { BalanceFunction, BalancePallet } from './balance.constants';
import {
  AssetMinBalanceConfig,
  AssetsBalanceConfig,
  SystemBalanceConfig,
  TokensBalanceConfig,
} from './balance.interfaces';

/* eslint-disable @typescript-eslint/no-use-before-define */
export function createBalanceBuilder<Asset>() {
  return {
    assets,
    min,
    system,
    tokens: (asset: number | Asset | 'MOVR' | 'KUSD' | 'AUSD') =>
      tokens<Asset>(asset),
  };
}

function assets(asset: number): AssetsBalanceConfig {
  return {
    pallet: BalancePallet.Assets,
    function: BalanceFunction.Account,
    path: ['balance'],
    getParams: (account: string) => [asset, account],
  };
}

function min(asset: number): AssetMinBalanceConfig {
  return {
    pallet: BalancePallet.Assets,
    function: BalanceFunction.Asset,
    path: ['minBalance'],
    getParams: () => [asset],
  };
}

function system(): SystemBalanceConfig {
  return {
    pallet: BalancePallet.System,
    function: BalanceFunction.Account,
    path: ['data'],
    getParams: (account: string) => [account],
    calc: ({ free, miscFrozen }: PalletBalancesAccountData) =>
      BigInt(free.sub(miscFrozen).toString()),
  };
}

function tokens<Asset>(
  asset: number | Asset | 'MOVR' | 'KUSD' | 'AUSD',
): TokensBalanceConfig<Asset> {
  return {
    pallet: BalancePallet.Tokens,
    function: BalanceFunction.Accounts,
    path: ['free'],
    getParams: (account: string) => [
      account,
      Number.isInteger(asset)
        ? { ForeignAsset: asset as number }
        : { Token: asset as Asset },
    ],
  };
}
