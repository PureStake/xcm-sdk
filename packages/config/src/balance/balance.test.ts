import { createBalanceBuilder } from './balance';

describe('Balance', () => {
  const account = '<ACCOUNT>';
  const balance = createBalanceBuilder();

  describe('assets', () => {
    const cfg = balance.assets(1);

    it('should return assets balance config', () => {
      expect(cfg).toMatchSnapshot();
    });

    it('should return assets balance params', () => {
      expect(cfg.getParams(account)).toMatchSnapshot();
    });
  });

  describe('min', () => {
    const cfg = balance.min(1);

    it('should return min balance config', () => {
      expect(cfg).toMatchSnapshot();
    });

    it('should return min balance config params', () => {
      expect(cfg.getParams()).toMatchSnapshot();
    });
  });

  describe('system', () => {
    const cfg = balance.system();

    it('should return system balance config', () => {
      expect(cfg).toMatchSnapshot();
    });

    it('should return system balance config params', () => {
      expect(cfg.getParams(account)).toMatchSnapshot();
    });
  });

  describe('tokens', () => {
    const cfg = balance.tokens('MOVR');

    it('should return tokens balance config', () => {
      expect(cfg).toMatchSnapshot();
    });

    it('should return tokens balance config with Token params', () => {
      expect(cfg.getParams(account)).toMatchSnapshot();
    });

    it('should return tokens balance config with ForeignAsset params', () => {
      const cfg2 = balance.tokens(5);

      expect(cfg2.getParams(account)).toMatchSnapshot();
    });
  });
});
