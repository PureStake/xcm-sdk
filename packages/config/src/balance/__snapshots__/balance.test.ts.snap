// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Balance assets should return assets balance config 1`] = `
Object {
  "function": "account",
  "getParams": [Function],
  "pallet": "assets",
  "path": Array [
    "balance",
  ],
}
`;

exports[`Balance assets should return assets balance params 1`] = `
Array [
  1,
  "<ACCOUNT>",
]
`;

exports[`Balance min should return min balance config 1`] = `
Object {
  "function": "asset",
  "getParams": [Function],
  "pallet": "assets",
  "path": Array [
    "minBalance",
  ],
}
`;

exports[`Balance min should return min balance config params 1`] = `
Array [
  1,
]
`;

exports[`Balance system should return system balance config 1`] = `
Object {
  "calc": [Function],
  "function": "account",
  "getParams": [Function],
  "pallet": "system",
  "path": Array [
    "data",
  ],
}
`;

exports[`Balance system should return system balance config params 1`] = `
Array [
  "<ACCOUNT>",
]
`;

exports[`Balance tokens should return tokens balance config 1`] = `
Object {
  "calc": [Function],
  "function": "accounts",
  "getParams": [Function],
  "pallet": "tokens",
}
`;

exports[`Balance tokens should return tokens balance config with ForeignAsset params 1`] = `
Array [
  "<ACCOUNT>",
  Object {
    "ForeignAsset": 5,
  },
]
`;

exports[`Balance tokens should return tokens balance config with Token params 1`] = `
Array [
  "<ACCOUNT>",
  Object {
    "Token": "MOVR",
  },
]
`;
